pyccel/stdlib/ndarrays/ndarrays.h

```diff
@@ -0,0 +1,68 @@
+#ifndef NDARRAYS_H
+# define NDARRAYS_H
+
+# include <stdlib.h>
+# include <complex.h>
+# include <string.h>
+# include <stdio.h>
+# include <stdarg.h>
+# include <stdbool.h> 
+typedef struct  s_slice_data
+{
+    int start;
+    int end;
+    int step;
+}               t_slice;
+
+enum e_types
+{
+        nd_int,
+        nd_float,
+        nd_double,
+        nd_cdouble
+};
+
+typedef struct  s_ndarray
+{
+    /* raw data buffer*/
+    union {
+            char            *raw_data;
+            int             *nd_int;
+            float           *nd_float;
+            double          *nd_double;
+            double complex  *nd_cdouble;
+            };
+    /* number of dimmensions */
+    int             nd;
+    /* shape 'size of each dimmension' */
+    int             *shape;
+    /* strides 'number of bytes to skip to get the next element' */
+    int             *strides;
+    /* type of the array elements */
+    enum e_types    type;
+    int             type_size;
+    int             length;
+    int             buffer_size;
+    bool            is_slice;
+}               t_ndarray;
+
+/* functions prototypes */
+
+/* allocations */
+t_ndarray   array_create(int nd, int *shape, enum e_types type);
+t_ndarray   array_init(char *temp, int nd, int *shape, enum e_types type);
+t_ndarray   array_ones(int nd, int *shape, enum e_types type);
+t_ndarray   array_zeros(int nd, int *shape, enum e_types type);

```

Comment:
Given that we are in the process of adding an `Allocate` node to the Pyccel AST, it seems preferable to have `array_init`, `array_ones` and `array_zeros` only initialize the array, without allocating it.

@saidctb @EmilyBourne What do you think?

Reply:
I'm not sure I agree. While the addition of the `Allocate` node will make it very easy to print:
```
x = array_create(...)
array_init(x, ...)
```
we usually try to have code that resembles the python code, unless the changes make the result more natural in the target language. I'm not sure that this is the case here.

If we did do this though, then the initialisation is potentially simple enough to not require a function.
I would probably agree that:
```
x = array_create(...)
memcpy(arr.raw_data, temp, arr.buffer_size);
```
looks more like standard c code.

Reply:
Now that I think about it, it makes a lot of sense to leave the initialization to the C printer... In many practical cases we will need some non-trivial initialization anyway. If we take this path, the simple initialization functions provided here would be only used in the unit tests.

Reply:
we need an empty which is for the array creation, and a function for filling the array with a value to use it for `np.full`, and all the other will use a combination of the two

Reply:
`empty` is equivalent to `array_create` but I agree that we should use the same naming convention

`array_ones` can use the full method, but I think we should keep `array_zeros` as you can allocate 0 much faster with memset than with a loop

Reply:
for np.full we either need an iterator function to use it on the sliced arrays or we just use the C printer to create the needed loops (from the C printer we can manage the different types without using the _generic macro)

Reply:
`full` creates an array so you don't need to worry about slices. That means you can just iterate over the flattened array

Reply:
is it better to write it in the C printer or create a generic function for it which mean we need a full function for each type the call them with the _generic macro

Reply:
After PR #521 the nodes `NumpyFull`, `NumpyEmpty`, `NumpyZeros`, `NumpyOnes`, etc. will not take care of memory allocation any more, because an `Allocate` node (which we can rename however we want) will be created at the semantic stage. So I hardly see the need for more than the C functions for allocation and deallocation, and maybe a few utility functions later on to help with the loops.

This discussion is becoming a bit confusing to me. Are you guys reconsidering PR #521 altogether?

Reply:
we wanted to make it compatible with the PR #521 which means we need to separate between array allocation and initialization , in here we have `array_ones` and `array_zeros` instead we can just add a`array_full` which initializes the array to whatever value you give it  with  the exception of array_zeros but the rest should use `array_full`, so then the allocation node in PR #521 will use `array_create`

Reply:
OK, then I suggest that we change the names as follows: `array_create` creates a new multi-dimensional array, while `array_fill` just fills it with a scalar value.
Any further optimization can be done at the printing stage.